File: .tmp_py4013/batch_04_sources/2478_number_of_beautiful_partitions.js
// LeetCode 2478 - Number of Beautiful Partitions
// https://leetcode.com/problems/number-of-beautiful-partitions/

/**
 * @param {string} s
 * @param {number} k
 * @param {number} minLength
 * @return {number}
 */
var beautifulPartitions = function(s, k, minLength) {
    const mod = 1000000007;
    const isPrime = (c) => c === '2' || c === '3' || c === '5' || c === '7';
    const n = s.length;
    if (!isPrime(s[0]) || isPrime(s[n - 1])) return 0;
    const dp = Array.from({ length: k + 1 }, () => Array(n + 1).fill(0));
    dp[0][0] = 1;
    for (let p = 1; p <= k; p++) {
        let pref = 0, j = 0;
        for (let i = 1; i <= n; i++) {
            while (j <= i - minLength) {
                if (j === 0 || (isPrime(s[j]) && !isPrime(s[j - 1]))) {
                    pref = (pref + dp[p - 1][j]) % mod;
                }
                j++;
            }
            if (!isPrime(s[i - 1])) dp[p][i] = pref;
        }
    }
    return dp[k][n];
};

File: .tmp_rs4013/rust_check_batch08/2478_number_of_beautiful_partitions.rs
struct Solution;
// LeetCode 2478 - Number of Beautiful Partitions
// https://leetcode.com/problems/number-of-beautiful-partitions/

impl Solution {
    pub fn beautiful_partitions(s: String, k: i32, min_length: i32) -> i32 {
        const MOD: i32 = 1_000_000_007;
        fn is_prime(c: u8) -> bool {
            matches!(c, b'2' | b'3' | b'5' | b'7')
        }
        let b = s.as_bytes();
        let n = b.len();
        if !is_prime(b[0]) || is_prime(b[n - 1]) {
            return 0;
        }
        let k = k as usize;
        let min_length = min_length as usize;
        let mut dp = vec![vec![0; n + 1]; k + 1];
        dp[0][0] = 1;
        for p in 1..=k {
            let mut pref = 0;
            let mut j = 0;
            for i in 1..=n {
                while j + min_length <= i {
                    if j == 0 || (is_prime(b[j]) && !is_prime(b[j - 1])) {
                        pref = (pref + dp[p - 1][j]) % MOD;
                    }
                    j += 1;
                }
                if !is_prime(b[i - 1]) {
                    dp[p][i] = pref;
                }
            }
        }
        dp[k][n]
    }
}

fn main() {}
